pure-wasm/src/lib.rs

```diff
@@ -113,43 +119,60 @@ pub extern "C" fn dealloc_response_buffer(ptr: *mut u8, len: usize) {
     }
 }
 
-// This is only used as a thread local so it's thread safe. If this gets moved out of the
-// thread local this will ruin someone's day. So don't do that. Please bro
-pub fn build_response(enabled: Option<bool>, error: Option<&str>) -> Vec<u8> {
-    BUILDER.with(|cell| {
-        let mut builder = cell.borrow_mut();
-        builder.reset();
+pub trait FlatbufferSerializable<TInput>: Follow<'static> + Sized {
+    fn as_flat_buffer(builder: &mut FlatBufferBuilder<'static>, input: TInput) -> WIPOffset<Self>;
 
-        let error_offset = error.map(|e| builder.create_string(e));
+    fn build_response(input: TInput) -> Vec<u8> {
+        BUILDER.with(|cell| {
+            let mut builder = cell.borrow_mut();
+            builder.reset();
 
-        let response = {
-            let mut resp_builder = ResponseBuilder::new(&mut builder);
-            if let Some(flag) = enabled {
-                resp_builder.add_enabled(flag);
-                resp_builder.add_has_enabled(true);
-            }
-            if let Some(err) = error_offset {
-                resp_builder.add_error(err);
-            }
-            resp_builder.finish()
-        };
+            let offset = Self::as_flat_buffer(&mut builder, input);
```

Comment:
We use this top level implemented function on the trait, which dispatches to the user defined trait impl. This means the surrounding lines of code here can't be forgotten so long as you follow the pattern